Adjust trimmed_pod_id and replace '.' with '-'

kd2718 opened the pull request:
closes: #16611 
related: #16980

Having a period in the name could cause issues with the hostname. Reverting to hyphen will should fix the issues. I also included changes talked about in a [previous PR](https://github.com/apache/airflow/pull/16980#pullrequestreview-721302569).

kd2718 committed:
replaced '.' with '-' and adjusted trimmed_pod_id per ticket comments

uranusjr commented on airflow/kubernetes/pod_generator.py:
```diff
@@ -460,9 +460,9 @@ def make_unique_pod_id(pod_id: str) -> str:
         safe_uuid = uuid.uuid4().hex  # safe uuid will always be less than 63 chars
         # Strip trailing '-' and '.' as they can't be followed by '.'
-        trimmed_pod_id = pod_id[:MAX_LABEL_LEN].rstrip('-.')
+        trimmed_pod_id = pod_id[:MAX_LABEL_LEN-len(safe_uuid)].rstrip('-.')
```
Please also update the comment above to explain the logic. It's also probably better to split this into two lines

```suggestion
        label_prefix_length = MAX_LABEL_LEN - len(safe_uuid)
        trimmed_pod_id = pod_id[:label_prefix_length].rstrip('-.')
```

Because you probably won't like how Black formats this in a single line 🙂

kd2718 replied:
thanks. I have updated with these changes

uranusjr reviewed (commented):
cc @dimon222 to take a look if possible

kd2718 committed:
added changes for PR

kd2718 committed:
fixed typo and updated tests to use new seperator. Some tests may still fail... investigating

uranusjr commented on airflow/kubernetes/pod_generator.py:
```diff
@@ -459,10 +459,14 @@ def make_unique_pod_id(pod_id: str) -> str:
-        safe_pod_id = f"{trimmed_pod_id}.{safe_uuid}"
+        # Get prefix length after subtracting the uuid length. Clean up . and - from
+        # end of podID because they can't be followed by '.'
+        label_prefix_length = MAX_LABEL_LEN - len(safe_uuid) - 1 # -1 for seperator
```
```suggestion
        label_prefix_length = MAX_LABEL_LEN - len(safe_uuid) - 1 # -1 for separator
```

kd2718 committed:
Update airflow/kubernetes/pod_generator.py

bad spelling

Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>

kd2718 committed:
added space for proper linting

kd2718 committed:
removed mock that was not being used

I added a mock, but it wasn't needed. I think it was also causing some test failure

kd2718 committed:
fixed more tests

kd2718 committed:
finally passing all tests locally

kd2718 committed:
fixed a few formatting errors

jedcunningham commented on tests/kubernetes/test_pod_generator.py:
```diff
@@ -658,8 +658,8 @@ def test_deserialize_model_file(self):
+        if len(pod_id) <= 63 - 33:
             assert len(parts[0]) == len(pod_id)
         else:
             assert len(parts[0]) <= 63
```
```suggestion
        parts = name.split("-")
        assert parts[0] == pod_id[:30]
```

Also the following line I can't suggest, line 666, no longer makes sense imo. We should check it against a fixed length, I think this is right:

```
assert len(parts[1]) == 32
```

kd2718 replied:
This makes much more sense. thanks!

jedcunningham commented on tests/kubernetes/test_pod_generator.py:
```diff
@@ -497,7 +497,7 @@ def test_ensure_max_label_length(self, mock_uuid):
         )
 
-        assert result.metadata.name == 'a' * 63 + '.' + self.static_uuid.hex
+        assert result.metadata.name == ('a' * 63)[:-33] + '-' + self.static_uuid.hex
```
```suggestion
        assert result.metadata.name == 'a' * 30 + '-' + self.static_uuid.hex
```

jedcunningham commented on tests/kubernetes/test_pod_generator.py:
```diff
@@ -684,7 +684,7 @@ def test_pod_name_is_valid(self, pod_id, expected_starts_with):
         ), "pod_id is invalid - fails allowed regex check"
 
-        assert name.rsplit(".")[0] == expected_starts_with
+        assert name[:-33] == expected_starts_with
```
```suggestion
        assert name.rsplit("-")[0] == expected_starts_with
```

Nit, I think the `rsplit` approach is easier to grok?

kd2718 replied:
My only worry is what if we change the "-" back to some other character... But then you could say the same worry about if we change the max pod length.

kd2718 replied:
Actually, this does create some issues. The test pod names are "pod-name-with-hyphen", so `name.rsplit("-")[0]` will just return `pod`. I switched it to use `"-".join(name.rsplit("-")[:-1])`. Unless we want to move back to the arbitrary `[:-33]` then this might be a better option.

kd2718 replied:
ref: https://github.com/apache/airflow/pull/19036/files#diff-60bb2e2d34655edf905c00b495e1feee30aac430040bfca7995c812841c84916R689

uranusjr replied:
> The test pod names are "pod-name-with-hyphen", so `name.rsplit("-")[0]` will just return `pod`.

Would `name.rsplit("-", 1)[0]` or `name.rpartition("-")[0]` work?

kd2718 replied:
`rsplit("-",1)[0]` works. Thanks for the tip. I am not as familiar with `rsplit`

jedcunningham commented on airflow/kubernetes/pod_generator.py:
```diff
@@ -459,10 +459,14 @@ def make_unique_pod_id(pod_id: str) -> str:
 
-        safe_pod_id = f"{trimmed_pod_id}.{safe_uuid}"
+        # Get prefix length after subtracting the uuid length. Clean up . and - from
+        # end of podID because they can't be followed by '.'
```
```suggestion
        # Get prefix length after subtracting the uuid length. Clean up '.' and '-' from
        # end of podID ('.' can't be followed by '-').
```

Technically we no longer need to remove any trailing`-`, but I think it makes sense to continue doing so.

kd2718 committed:
:qUpdate tests/kubernetes/test_pod_generator.py

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

kd2718 committed:
Update tests/kubernetes/test_pod_generator.py

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

kd2718 committed:
Update tests/kubernetes/test_pod_generator.py

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

kd2718 committed:
Update airflow/kubernetes/pod_generator.py

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

kd2718 committed:
fixed test max length from github comments

kd2718 committed:
removed join and replaced with rsplit(-,1)

uranusjr commented on tests/kubernetes/test_pod_generator.py:
```diff
@@ -658,11 +658,12 @@ def test_deserialize_model_file(self):
-        parts = name.split(".")
-        if len(pod_id) <= 63:
+        parts = name.split("-")
+        if len(pod_id) <= 63 - 33:
```
Maybe add a comment to explain why 63 and 33 are significant?

kd2718 replied:
I added a few notes explaining where these numbers came from.

uranusjr reviewed (approved):
Mostly ready but one nitpick

jedcunningham commented on tests/kubernetes/test_pod_generator.py:
```diff
@@ -658,11 +658,12 @@ def test_deserialize_model_file(self):
         else:
             assert len(parts[0]) <= 63
+        assert len(parts[1]) == 32
         assert len(parts[1]) <= 63
```
```suggestion
```

Don't need to check it's <=63 when we've just asserted it's 32.

kd2718 replied:
good catch

jedcunningham commented on tests/kubernetes/test_pod_generator.py:
```diff
@@ -658,11 +658,12 @@ def test_deserialize_model_file(self):
+        if len(pod_id) <= 63 - 33:
             assert len(parts[0]) == len(pod_id)
         else:
             assert len(parts[0]) <= 63
```
```suggestion
        assert parts[0] == pod_id[:30]
```

Did this get lost in the shuffle?

kd2718 replied:
chagned to `assert len(parts[0]) <= 30`. The prefix can be less than 30, but not larger.

kd2718 committed:
cleaned up test_pod_name_confirm_to_max_length as noted by comments

kd2718 committed:
fixed spelling error on separating

jedcunningham commented on airflow/kubernetes/pod_generator.py:
```diff
@@ -459,10 +459,14 @@ def make_unique_pod_id(pod_id: str) -> str:
+
+        # previously used a '.' as the separator, but this could create errors in some situations
+        safe_pod_id = f"{trimmed_pod_id}-{safe_uuid}"
         return safe_pod_id
```
```suggestion
        return f"{trimmed_pod_id}-{safe_uuid}"
```

nit

kd2718 committed:
returned directly from function rather than creating a new variable

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>
